crates/providers/src/layers/nonce.rs

```diff
@@ -0,0 +1,204 @@
+use crate::{PendingTransaction, Provider, ProviderLayer};
+use alloy_network::{Network, TransactionBuilder};
+use alloy_primitives::{Address, B256, U64};
+use alloy_rpc_client::{ClientRef, WeakClient};
+use alloy_transport::{Transport, TransportResult};
+use async_trait::async_trait;
+use dashmap::DashMap;
+use std::{
+    marker::PhantomData,
+    sync::atomic::{AtomicU64, Ordering},
+};
+
+/// A layer that fills nonces on transactions.
+///
+/// The layer will fetch the transaction count for any new account it sees, store it locally and
+/// increment the locally stored nonce as transactions are sent via [`Provider::send_transaction`].
+///
+/// If you use layers that redirect the behavior of [`Provider::send_transaction`] (e.g.
+/// [`SignerLayer`]), you should add this layer before those.
+///
+/// # Note
+///
+/// If the transaction request does not have a sender set, this layer will not fill nonces.
+///
+/// # Example
+///
+/// ```rs
+/// # async fn test<T: Transport + Clone, S: NetworkSigner<Ethereum>>(transport: T, signer: S) {
+/// let provider = ProviderBuilder::<_, Ethereum>::new()
+///     .layer(ManagedNonceLayer)
+///     .signer(EthereumSigner::from(signer)) // note the order!
+///     .network::<Ethereum>()
+///     .provider(RootProvider::new(transport));
+///
+/// provider.send_transaction(TransactionRequest::default()).await;
+/// # }
+/// ```
+///
+/// [`SignerLayer`]: crate::layers::SignerLayer
+pub struct ManagedNonceLayer;
+
+impl<P, N, T> ProviderLayer<P, N, T> for ManagedNonceLayer
+where
+    P: Provider<N, T>,
+    N: Network,
+    T: Transport + Clone,
+{
+    type Provider = ManagedNonceProvider<N, T, P>;
+
+    fn layer(&self, inner: P) -> Self::Provider {
+        ManagedNonceProvider { inner, nonces: DashMap::default(), _phantom: PhantomData }
+    }
+}
+
+/// A provider that manages account nonces.
+///
+/// Fills nonces for transaction requests if unset.
+///
+/// # Note
+///
+/// If the transaction requests do not have a sender set, this provider will not set nonces.
+///
+/// You cannot construct this provider directly. Use [`ProviderBuilder`] with a
+/// [`ManagedNonceLayer`].
+///
+/// [`ProviderBuilder`]: crate::ProviderBuilder
+pub struct ManagedNonceProvider<N, T, P>
+where
+    N: Network,
+    T: Transport + Clone,
+    P: Provider<N, T>,
+{
+    inner: P,
+    nonces: DashMap<Address, AtomicU64>,
+    _phantom: PhantomData<(N, T)>,
+}
+
+impl<N, T, P> ManagedNonceProvider<N, T, P>
+where
+    N: Network,
+    T: Transport + Clone,
+    P: Provider<N, T>,
+{
+    async fn get_nonce(&self, from: Address) -> TransportResult<U64> {
```

Comment:
this could potentially be problematic, because now once inserted into the map, the nonce is fixed

if you send two tx without a nonce sequentially through this provider, the second one will reuse the previous nonce

Reply:
that's not the case, as you can see in the test